init.lua

```diff
@@ -56,20 +36,80 @@ local function re_run_failed_jobs(full_repo, run_id)
         verify_peer = false,
     }
     local res = client:request('POST', url, '', op)
-    log.info('Api call for re-running '..run_id..' job finished with status '..res.status)
+    log.info('Api call for re-running '..run_id..' run finished with status '..res.status)
     log.info('Api response body is '..res.body)
 end
 
+
+--- Handle an incoming webhook from GitHub
+-- @see https://docs.github.com/en/developers/webhooks-and-events/webhooks/about-webhooks
 function webhook_handler(req)
-    local job = req:json()
-    local run_id = job.workflow_job.run_id
-    local full_repo = job.repository.full_name
-    if is_job_failed(job) and needs_restart(run_id) then
-        fiber.create(function() re_run_failed_jobs(full_repo, run_id) end)
+    local payload = req:json()
+    if payload.workflow_job == nil then
+        log.error('Empty payload')
+        return { status = 204 }
+    end
+    if payload.workflow_job.run_attempt >= 4 then
+        log.info('Attempt >= 4, no action needed')
+        return { status = 204 }
+    end
+    local workflow_run_id = payload.workflow_job.run_id
+    local full_repo = payload.repository.full_name
+
+    describe(payload)
+
+    if payload.action == 'queued' then
+        local workflow_record = box.space.jobs:get({ workflow_run_id })
+        if workflow_record then
+            box.space.jobs:update(workflow_run_id, { { '+', 'count', 1}})
+        else
+            box.space.jobs:insert({ workflow_run_id, 1, true})
+        end
+        workflow_record = box.space.jobs:get({ workflow_run_id })
+        log.info('Job queued in workflow '..workflow_run_id..': '..tostring(workflow_record))
+        return { status = 200 }
+    elseif payload.action == 'in_progress' then
+        return { status = 200 }
+    elseif payload.action == 'completed' then
+        local workflow_record = box.space.jobs:get({ workflow_run_id })
+        if workflow_record then
+            box.space.jobs:update(workflow_run_id, { { '-', 'count', 1}})
+            if payload.workflow_job.conclusion == 'failure' then
+                box.space.jobs:update(workflow_run_id, { { '=', 'fixed', false}})
+            end
+
+            workflow_record = box.space.jobs:get({ workflow_run_id })
+            log.info('Completed job in workflow '..workflow_run_id..': '..tostring(workflow_record))
+            if workflow_record.count == 0 and workflow_record.fixed == false then
+                re_run_failed_workflow(full_repo, workflow_run_id)
+                log.info('Rerunning workflow with run_id '..workflow_run_id..
+                        ' the current count of jobs in matrix is '..workflow_record.count..
+                        ' and fixed equal to '..tostring(workflow_record.fixed))
+
+            end
+
+            if workflow_record.count == 0 then
+                box.space.jobs:delete({ workflow_run_id })
+            end
+            return { status = 201 }
```

Comment:
In this block the code should return either 200 or 201 depending on whether it restarted the workflow.